datalake/src/main/java/com/sequenceiq/datalake/service/resize/recovery/ResizeRecoveryService.java

```diff
@@ -0,0 +1,116 @@
+import com.sequenceiq.flow.domain.FlowLog;
+import com.sequenceiq.flow.service.flowlog.FlowChainLogService;
+import com.sequenceiq.sdx.api.model.SdxRecoverableResponse;
+import com.sequenceiq.sdx.api.model.SdxRecoveryRequest;
+import com.sequenceiq.sdx.api.model.SdxRecoveryResponse;
+
+@Component
+/**
+ * Provides entrypoint for recovery of failed SDX resize.
+ *
+ * The main entry point is {@code triggerRecovery}, which starts a cloudbreak Flow to recover the Data Lake.
+ * To ensure a Resize recovery is appropriate, use one of {@code canRecover} or {@code validateRecovery}.
+ */
+public class ResizeRecoveryService implements RecoveryService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResizeRecoveryService.class);
+
+    @Inject
+    private SdxService sdxService;
+
+    @Inject
+    private SdxStatusService sdxStatusService;
+
+    @Inject
+    private SdxReactorFlowManager sdxReactorFlowManager;
+
+    @Inject
+    private Flow2Handler flow2Handler;
+
+    @Inject
+    private FlowChainLogService flowChainLogService;
+
+    /**
+     * Determines if it's possible to run a <em>Resize</em> recovery, returning a simple boolean indicator
+     *
+     * @param crn       the CRN of the user requesting recovery
+     * @param nameOrCrn the name or CRN of the resource to recover
+     * @return true if resize recovery is possible.
+     */
+    public boolean canRecover(String crn, NameOrCrn nameOrCrn) {
+        SdxCluster cluster = sdxService.getByNameOrCrn(crn, nameOrCrn);
+        FlowLog flowLog = flow2Handler.getFirstStateLogfromLatestFlow(cluster.getId());
+        if (DatalakeResizeFlowEventChainFactory.class.getSimpleName().equals(flowChainLogService.getFlowChainType(flowLog.getFlowChainId()))) {
+            SdxStatusEntity actualStatusForSdx = getSdxStatusEntity(crn, nameOrCrn);
+            if (actualStatusForSdx.getStatus().equals(DatalakeStatusEnum.STOP_FAILED)) {
+
+                return true;
+            }
+
+        }
+        return false;
+    }
+
+    private SdxStatusEntity getSdxStatusEntity(String crn, NameOrCrn nameOrCrn) {
+        SdxCluster cluster = sdxService.getByNameOrCrn(crn, nameOrCrn);
+        return sdxStatusService.getActualStatusForSdx(cluster);
+    }
+
+    /**
+     * Determines if it's possible to run a <em>Resize</em> recovery, returning a detailed validation message.
+     * <p>
+     * This is distinguished from {@code canRecover} by including a detailed validation message, usable as a web response.
+     *
+     * @param crn       the CRN of the user requesting recovery
+     * @param nameOrCrn the name or CRN of the resource to recover
+     * @return detailed validation message, as a web response
+     */
+    public SdxRecoverableResponse validateRecovery(String crn, NameOrCrn nameOrCrn) {
+        SdxStatusEntity actualStatusForSdx = getSdxStatusEntity(crn, nameOrCrn);
+        if (actualStatusForSdx.getStatus().equals(DatalakeStatusEnum.STOP_FAILED)) {
```

Comment:
Did we specify somewhere why the DL being in a STOP_FAILED state indicates recovery is possible?
I ask since I assumed there were a few other cases where we wanted recovery to be available:

- Creation of the new DL failing (which would probably leave the "primary" DL in a FAILED state)
- Restore failing (not sure what state this would end up in)
- Possibly even other steps failing like Detach, etc. which would leave the DL in a STOPPED state

Reply:
during a resize, a Datalake only enters the STOP_FAILED state when the Stop Action fails. This doesn't cover any other failure a resize might encounter to keep the scope of this change small.